File: neetcode150/18_bit_manipulation/0007_reverse_integer.cpp
/*
 * =====================================================================
 * LeetCode 7  : Reverse Integer                                    (Medium)
 * https://leetcode.com/problems/reverse-integer/
 * Category   : Bit manipulation / integer arithmetic
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given a signed 32-bit integer x, return x with its digits reversed.
 *   If reversing causes the value to fall outside [-2^31, 2^31-1],
 *   return 0.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Repeatedly peel off the last digit (x % 10), append it to a running
 *   answer (rev = rev*10 + d), and truncate x (x /= 10). The catch is
 *   32-bit overflow: building rev in 64 bits lets us test the range
 *   *after* the reversal, then clamp.
 * ---------------------------------------------------------------------
 * APPROACH 1 · Arithmetic pop/append (best)
 *   1. Keep the sign of x, work on the absolute value.
 *   2. Loop: d = x % 10 ; rev = rev*10 + d ; x /= 10.
 *   3. Apply sign, then test 32-bit range; return 0 if out of range.
 * APPROACH 2 · String reverse with built-in overflow guard
 *   1. Convert |x| to a string, reverse it, parse back to int.
 *   2. Apply sign, then clamp to 32-bit.
 * ---------------------------------------------------------------------
 * DEEP DIVE · WHY arithmetic beats string here
 *   Modulo 10 is a single x86 'idiv' remainder; integer multiply by 10
 *   and add is one IMUL + one ADD — no heap allocation, no parsing
 *   overhead, no UTF-8 decode. The 32-bit bound is INT32_MAX = 2147483647
 *   and INT32_MIN = -2147483648; building the result in a 64-bit long
 *   guarantees the comparison is always well defined (no UB from
 *   signed overflow during the build itself). Also note: 64-bit build
 *   handles INT32_MIN safely — INT32_MIN / -1 overflows in 32-bit, but
 *   never in 64-bit.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(log10 |x|) = O(1)  — at most 10 digits.
 * MEMORY COMPLEXITY: O(1)                  — rev fits in a long (8 bytes).
 * ---------------------------------------------------------------------
 */
#include <iostream>
#include <string>
#include <algorithm>     // std::reverse
#include <cstdint>
#include <chrono>
#include <sys/resource.h>

/* ---------------- Approach 1: arithmetic pop/append ---------------- */
class Solution_1 {
public:
    int reverse(int x) {
        long long rev = 0;                 // 64-bit: never overflows mid-build
        long long n = x;                   // copy so % and /= work on magnitude
        while (n != 0) {
            int d = static_cast<int>(n % 10);  // last digit (sign preserved)
            rev = rev * 10 + d;                // append to the growing answer
            n /= 10;                           // drop the digit we consumed
        }
        if (rev < INT32_MIN || rev > INT32_MAX) return 0;  // clamp to 32-bit
        return static_cast<int>(rev);
    }
};

/* ---------------- Approach 2: string reverse ---------------- */
class Solution_2 {
public:
    int reverse(int x) {
        long long n = x;                          // magnitude-safe copy
        bool neg = n < 0;
        std::string s = std::to_string(neg ? -n : n);  // digits only
        std::reverse(s.begin(), s.end());         // flip the string
        long long rev = std::stoll(s);            // parse back
        if (neg) rev = -rev;                      // restore sign
        if (rev < INT32_MIN || rev > INT32_MAX) return 0;
        return static_cast<int>(rev);
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() {
    struct rusage r{};
    getrusage(RUSAGE_SELF, &r);
    return r.ru_maxrss;  // kilobytes on Linux
}

int main() {
    struct Case { int x; int expect; };
    const Case cases[] = {
        {123, 321}, {-123, -321}, {120, 21},
        {0, 0}, {1534236469, 0}, {-2147483648, 0}
    };
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        int r1 = Solution_1().reverse(c.x);
        int r2 = Solution_2().reverse(c.x);
        bool ok = r1 == c.expect && r2 == c.expect;
        allOk = allOk && ok;
        std::cout << "x=" << c.x << " → A1=" << r1 << " A2=" << r2
                  << " (want " << c.expect << ") " << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all 6 cases on both approaches" : "FAIL : some case wrong")
              << " | time: " << (t1 - t0) << " us | mem: " << (memKb() - mem0) << " KB\n";
    /* benchmark: 1e6 reversals of a pathological input */
    t0 = nowUs();
    for (int i = 0; i < 1000000; ++i) Solution_1().reverse(1534236469);
    t1 = nowUs();
    std::cout << "BENCH A1: " << (t1 - t0) << " us for 1e6 calls\n";
    t0 = nowUs();
    for (int i = 0; i < 1000000; ++i) Solution_2().reverse(1534236469);
    t1 = nowUs();
    std::cout << "BENCH A2: " << (t1 - t0) << " us for 1e6 calls\n";
    return 0;
}

File: neetcode150/18_bit_manipulation/0007_reverse_integer.py
"""
=====================================================================
LeetCode 7  : Reverse Integer                                    (Medium)
https://leetcode.com/problems/reverse-integer/
Category   : Bit Manipulation
---------------------------------------------------------------------
PROBLEM
    Given a signed 32-bit integer `x`, return `x` with its decimal
    digits reversed. If reversing would fall OUTSIDE the signed
    32-bit range [-2^31, 2^31 - 1], return 0.
    Examples: 123 → 321     -123 → -321     1534236469 → 0 (overflow)
---------------------------------------------------------------------
INTUITION
    Pop the last digit (x % 10) and push it onto the result
    (result = result * 10 + digit) — writing the number backwards
    by hand. In C++/Java the multiply can overflow, so guards run
    BEFORE the multiplication. Python's ints are UNBOUNDED, so it
    never overflows naturally — the guard is an EMULATION of the
    int32 hardware contract, which is exactly what the problem tests.
---------------------------------------------------------------------
APPROACH 1 — Pop & push with pre-multiply guards (BEST)
    Check `result > INT_MAX // 10` before growing (any growth
    overflows) or `result == INT_MAX // 10 and digit > 7` (only the
    largest digit overflows at the exact boundary). Mirror on the
    negative side with -214748364 and guard digit -8 — two's
    complement gives the negative range one extra unit, so the two
    guard digits DIFFER (7 vs -8).
    Time  Complexity : O(log10 |x|)   Space Complexity : O(1)
---------------------------------------------------------------------
APPROACH 2 — Unbounded accumulation, single final range check
    Exploit Python's big ints: accumulate WITHOUT guards (nothing can
    overflow), then clamp once at the end: if the reversal escapes
    [-2^31, 2^31 - 1] → 0. Same loop, fewer branches, and it makes
    the "mask to 32 bits" idea explicit: the hardware does this
    clamping silently, so the algorithm must do it deliberately.
    Time  Complexity : O(log10 |x|)   Space Complexity : O(1)
---------------------------------------------------------------------
REAL LIFE
    • Integer parsing in compilers and DB wire protocols: every
      atoi/strtol implementation carries exactly this pre-multiply
      guard so a hostile 20-digit string cannot wrap the value.
    • Network byte order (big-endian) conversion when reversing
      digit/byte streams between CPU architectures.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import List, Tuple

# 32-bit signed range boundaries (Python emulates these explicitly).
INT32_MAX = 2 ** 31 - 1
INT32_MIN = -2 ** 31


# =====================================================================
# APPROACH 1 : Pop & push with pre-multiply overflow guards (best)
# =====================================================================
class Solution_1:
    """
    Purpose : Reverse the decimal digits of a 32-bit signed int,
              returning 0 if the reversal leaves the int32 range.
    Inputs  : x — the integer whose digits must be reversed.
    Output  : int — reversed digits, or 0 on overflow.
    """

    def reverse(self, x: int) -> int:
        neg = x < 0
        n = -x if neg else x            # work on magnitude: // truncates
        result = 0

        # Peel digits off the right until nothing remains.
        while n != 0:
            digit = n % 10
            n //= 10

            # Positive guard: INT32_MAX // 10 = 214748364, guard digit 7.
            # If result already exceeds the quotient, ANY append grows
            # past the boundary; exactly at the quotient only digit > 7
            # overflows. The magnitude bound matches the 32-bit signed
            # range because |INT32_MIN| is only 1 wider than INT32_MAX,
            # and reversing -2147483648 overflows the magnitude anyway.
            if result > INT32_MAX // 10 or (result == INT32_MAX // 10 and digit > 7):
                return 0

            result = result * 10 + digit

        return -result if neg else result


# =====================================================================
# APPROACH 2 : Unbounded accumulation, single final range check
# =====================================================================
class Solution_2:
    """
    Purpose : Reverse the digits using Python's unbounded ints, then
              clamp to 0 with ONE range check at the end.
    Inputs  : x — the integer whose digits must be reversed.
    Output  : int — reversed digits, or 0 on overflow.
    """

    def reverse(self, x: int) -> int:
        # Python ints are arbitrary precision: the reversal of any
        # int32 (≤ 10 digits) can NEVER overflow, so no mid-loop
        # guards exist. The loop is just the raw digit shuffle.
        sign = -1 if x < 0 else 1
        value = abs(x)
        result = 0
        while value != 0:
            result = result * 10 + (value % 10)
            value //= 10
        result *= sign

        # ONE emulated int32 boundary check. In C++ the hardware
        # truncates silently here — returning 0 is the contract.
        if result < INT32_MIN or result > INT32_MAX:
            return 0
        return result


# =====================================================================
# BENCHMARK — run both approaches and compare time + peak memory
# =====================================================================
if __name__ == "__main__":
    # Test table: {input, expected} — includes an overflow case and
    # INT_MIN, the value whose negation still equals itself in int32.
    tests: List[Tuple[int, int]] = [
        (123, 321),
        (-123, -321),
        (1534236469, 0),
        (-2147483648, 0),
    ]

    # --- Approach 1: time + peak memory (tracemalloc snapshot) ---
    tracemalloc.start()
    t0 = time.perf_counter()
    ok_1 = True
    for inp, exp in tests:
        got = Solution_1().reverse(inp)
        pass_ = got == exp
        ok_1 = ok_1 and pass_
        print(f"  A1 reverse({inp}) = {got}  expected {exp}  {'PASS' if pass_ else 'FAIL'}")
    t1 = time.perf_counter()
    _, peak_1 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # --- Approach 2: time + peak memory ---
    tracemalloc.start()
    t0b = time.perf_counter()
    ok_2 = True
    for inp, exp in tests:
        got = Solution_2().reverse(inp)
        pass_ = got == exp
        ok_2 = ok_2 and pass_
        print(f"  A2 reverse({inp}) = {got}  expected {exp}  {'PASS' if pass_ else 'FAIL'}")
    t1b = time.perf_counter()
    _, peak_2 = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # Summary lines: time in µs, peak memory in KB.
    print(f"Approach 1 (guarded pop/push)   : time = {(t1 - t0) * 1e6:.2f} µs  peak-mem = {peak_1 / 1024:.2f} KB")
    print(f"Approach 2 (unbounded + clamp)  : time = {(t1b - t0b) * 1e6:.2f} µs  peak-mem = {peak_2 / 1024:.2f} KB")

    print("PASS : both approaches agree with every expected value."
          if ok_1 and ok_2
          else "FAIL : at least one test mismatched.")
